### PR TITLE
Add `getStatistics()` to python API

### PR DESCRIPTION
**Pull request by nafur**
This PR adds `Solver.getStatistics()` to the python API. To make the usage a bit more pythonic, we do not expose the iterator interface of `api::Statistics` but instead offer `.get()` which returns the whole range as a dictionary. The `::get()` method to obtain a single statistic value is instead implemented via `__getitem__()`.
The PR also includes the corresponding unit tests.

### CODE REVIEW

File: src/api/cpp/cvc5.h [hunk 1]
```diff
@@ -2840,8 +2840,12 @@ class CVC5_EXPORT Stat
   friend std::ostream& operator<<(std::ostream& os, const Stat& sv);
   /** Representation of a histogram: maps names to frequencies. */
   using HistogramData = std::map<std::string, uint64_t>;
-  /** Can only be obtained from a `Statistics` object. */
-  Stat() = delete;
+  /**
+   * Create an empty statistics object. None of the ``isX()``and ``getX()``
```
Comment (yoni206) [inline src/api/cpp/cvc5.h:2844]: Perhaps add assertions in such functions?

File: src/api/cpp/cvc5.h [hunk 2]
```diff
@@ -2958,6 +2963,9 @@ class CVC5_EXPORT Statistics
     bool d_showDefault = false;
   };
 
+  /** Creates an empty statistics object. */
+  Statistics() = default;
```
Comment (yoni206) [inline src/api/cpp/cvc5.h:2967]: Why is this needed? Can you include an explanation in the documentation? It is clear why an empty constructor is needed, but not why it needs to be called from the cpp api.

File: src/api/python/cvc5.pxd [hunk 3]
```diff
@@ -380,6 +382,29 @@ cdef extern from "api/cpp/cvc5.h" namespace "cvc5::api":
...
+        double getDouble() except +
+        bint isString() except +
+        string getString() except +
+        bint isHistogram() except +
+        c_map[string,uint64_t] getHistogram() except +
+
+    cdef cppclass Statistics:
+        Statistics() except +
+        cppclass iterator:
+            iterator() except +
```
Comment (yoni206) [inline src/api/python/cvc5.pxd:400]: Is this needed?

File: src/api/cpp/cvc5.h (hunk 2 above)
Comment (nafur) [reply src/api/cpp/cvc5.h:2967]: A python statistics object needs to own the a `api::Statistics` objects. I haven't found a way to properly construct it (by copy-constructing from `Solver::getStatistics()`). Instead the `Statistics.cstats` member is first default-constructed and then copy-assigned to. Furthermore, this is also subject to visibility on the C++ side, so I can't make it private.

File: src/api/python/cvc5.pxd (hunk 3 above)
Comment (nafur) [reply src/api/python/cvc5.pxd:400]: What exactly do you mean? The whole `Statistics` class?

File: src/api/cpp/cvc5.h (hunk 1 above)
Comment (nafur) [reply src/api/cpp/cvc5.h:2844]: Yeah, I guess we should do that.

File: src/api/python/cvc5.pxd (hunk 3 above)
Comment (yoni206) [reply src/api/python/cvc5.pxd:400]: The iterator in the cython cppclass

Comment (nafur) [reply src/api/python/cvc5.pxd:400]: Yes: the list of statistics is exposed via `begin()` and `end()`.

**Review by yoni206 — APPROVED** [review]
